doc/man/gnunet-monkey.1
.TH GNUNET\-MONKEY 1 "Jan 4, 2012" "GNUnet"

.SH NAME
gnunet\-monkey \- automatically debugging for services

.SH SYNOPSIS
.B gnunet-monkey
.RI [ options ]
.br

.SH DESCRIPTION
\fBgnunet\-monkey\fP used to automatically debug a service.

.SH OPTIONS
.B
.IP "\-h, \-\-help"
Print short help on options.
.B
.IP "\-L LOGLEVEL, \-\-loglevel=LOGLEVEL"
Use LOGLEVEL for logging.  Valid values are DEBUG, INFO, WARNING and ERROR.
.B
.IP "\-m, \-\-mode"
Monkey's mode of operation: text: for writing debugging report in a text file, email: for sending debugging report to an email address.
.B
.IP "\-b, \-\-binary"
Path of the binary to debug with Monkey.
.B
.IP "\-o, \-\-output"
Only required in text mode: The path of the text file in which Monkey should write the debugging report.
.B
.IP "\-v, \-\-version"
Print GNUnet version number.


.SH BUGS
Report bugs by using Mantis <https://gnunet.org/mantis/> or by sending electronic mail to <gnunet\-developers@gnu.org>

.SH SEE ALSO
gnunet\-service\-monkey(1)


ChangeLog
Mon Oct 31 18:46:36 CET 2011
	Releasing GNUnet 0.9.0pre4.  This release includes a working
	implementation of mesh (unoptimized), a working implementation
	of the VPN (could use more testing), a working implementation 
	of the WLAN transport (still very slow, unoptimized) and
	many, many bugfixes all over the place.  The major known
	issues are:

	* TESTING is known to sometimes crash on shutdown
	* DV is known to crash / not work properly
	* CHAT is experimental and was not reviewed yet

	More details to individual bugs that were fixed in this
	release are documented at:
	https://gnunet.org/bugs/changelog_page.php?version_id=182

Mon Sep 19 14:16:22 CEST 2011
	Releasing GNUnet 0.9.0pre3.  Instead of listing all that has
	changed, here is the short list of what is known NOT to work,
	in order of minor flaws to major issues, by module:

	* FS still has a few missing features, but largely works
	* TRANSPORT ATS is not used for actual bandwidth assignment,
	          important options to make it truly useful are still missing
	* TESTING does not support incremental HELLO changes, 
	          code has known shutdown issues (can crash)
	* VPN is experimental, but has been shown to work to some degree
	* MESH does not work (mock API emulates for VPN)
        * TRANSPORT-WLAN is experimental
        * TRANSPORT-HTTP is experimental, has known bugs
	* TRANSPORT-DV is experimental, has known bugs, protocol may still change
	* CHAT is experimental, has known bugs and was not reviewed yet

Thu Dec 23 23:33:05 CET 2010
	Releasing GNUnet 0.9.0pre2.

Thu Oct	 7 20:51:05 CEST 2010
	Removed dependency on GNU MP (GMP). The code now uses the libgcrypt MPI
	interface.

Sat Jul  3 20:47:45 CEST 2010
	Releasing GNUnet 0.9.0pre1.

Wed Jun 23 16:34:38 CEST 2010
	Added support for systemd-compatible passing of listen-sockets
	by ARM to services as well as systemd compatibility for gnunet-service-arm
	itself.  At least for non-MINGW systems this should work.

Sat Jun  5 18:08:39 CEST 2010
	Added support for UNIX domain sockets, code also defaults to
	them when available.

Sun May  2 13:49:10 CEST 2010
	Fixed problem with platform-dependence of format for IP addresses
	in HELLOs for TCP and UDP transport.
	Reduced address length field in HELLOs to 16 bit (was 32 bit).
	These changes break transport compatibility.

Fri Apr 16 18:19:05 CEST 2010
	Nearly complete rewrite and new overall architecture.  Many
	features are still missing, but basic system seems to be
	working again.  Maybe time to again track major changes
	in the ChangeLog.  Releasing 0.9.0pre0.


Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent.
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).

Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in
	util/network/dns.c; improved IPv6 support for
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.


src/monkey/gnunet-monkey.c
/*
     This file is part of GNUnet.
     (C) 2010, 2011 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 3, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file monkey/gnunet-monkey.c
 * @brief Monkey: gnunet automated debugging tool
 */

#include <stdio.h>
#include "platform.h"
#include "gnunet_common.h"
#include "gnunet_getopt_lib.h"
#include "gnunet_program_lib.h"
#include "gnunet_monkey_action.h"

static const char *mode;
static const char *dumpFileName;
static const char *binaryName;
static const char *emailAddress;
static const char *edbFilePath;
static const char *gdbBinaryPath;
static const char *inspectExpression;
static const char *inspectFunction;
static int ret = 0;

/**
 * Main function that will launch the action api.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param c configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile, const struct GNUNET_CONFIGURATION_Handle *c)
{
  int result;
  struct GNUNET_MONKEY_ACTION_Context *cntxt;

  if (strcasecmp (mode, "email") == 0)
    {
      if (NULL == emailAddress)
	{
	  GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
		      "Working in email mode requires an email address!\n");
	  ret = 1;
	  return;
	}
    }
  else if (strcasecmp (mode, "text") == 0)
    {
      if (NULL == dumpFileName)
	{
	  GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
		      "Working in text mode requires a path for the dump file!\n");
	  ret = 1;
	  return;
	}
    }

  /* Initialize context for the Action API */
  cntxt = GNUNET_malloc (sizeof (struct GNUNET_MONKEY_ACTION_Context));
  cntxt->binary_name = binaryName;
  cntxt->expression_database_path = edbFilePath;
  cntxt->gdb_binary_path = gdbBinaryPath;
  cntxt->inspect_expression = inspectExpression;
  cntxt->inspect_function = inspectFunction;

  result = GNUNET_MONKEY_ACTION_rerun_with_gdb (cntxt);
  switch (result)
    {
      int retVal;
    case GDB_STATE_ERROR:
      break;
    case GDB_STATE_EXIT_NORMALLY:
      GNUNET_log (GNUNET_ERROR_TYPE_INFO,
		  "Debug with gdb, program exited normally!\n");
      /*FIXME: Valgrind should be launched here */
      break;
    case GDB_STATE_STOPPED:
      /*FIXME: Expression Database should be inspected here (before writing the report) */
      retVal = GNUNET_MONKEY_ACTION_inspect_expression_database (cntxt);
      if (GNUNET_NO == retVal)
	{
	  GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
		      "Error using Expression Database!\n");
	  ret = 1;
	  break;
	}
      else if (GDB_STATE_ERROR == retVal)
	{
	  /* GDB could not locate a NULL value expression, launch Valgrind */
	  retVal = GNUNET_MONKEY_ACTION_rerun_with_valgrind (cntxt);
	  if (GNUNET_NO == retVal)
	    {
	      GNUNET_log (GNUNET_ERROR_TYPE_ERROR, "Error using Valgrind!\n");
	      ret = 1;
	      break;
	    }
	}
      if (GNUNET_OK != GNUNET_MONKEY_ACTION_format_report (cntxt))
	{
	  GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
		      "Error in generating debug report!\n");
	  ret = 1;
	}
      if (strcasecmp (mode, "email") == 0)
	{
	  if (GNUNET_OK != GNUNET_MONKEY_ACTION_report_email (cntxt))
	    {
	      GNUNET_log (GNUNET_ERROR_TYPE_ERROR, "Error sending email!\n");
	      ret = 1;
	    }
	}
      else
	{
	  /* text mode */
	  if (GNUNET_OK !=
	      GNUNET_MONKEY_ACTION_report_file (cntxt, dumpFileName))
	    {
	      GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
			  "Error in saving debug file!\n");
	      ret = 1;
	    }
	}
      break;
    default:
      break;
    }
  GNUNET_MONKEY_ACTION_delete_context(cntxt);
}


int
main (int argc, char *argv[])
{
  static const struct GNUNET_GETOPT_CommandLineOption options[] = {
    {'m', "mode", NULL,
     gettext_noop
     ("monkey's mode of operation: options are \"text\" or \"email\""),
     GNUNET_YES, &GNUNET_GETOPT_set_string, &mode},
    {'b', "binary", NULL,
     gettext_noop ("binary for program to debug with monkey"),
     GNUNET_YES, &GNUNET_GETOPT_set_string, &binaryName},
    {'o', "output", NULL,
     gettext_noop
     ("path to file to dump monkey's output in case of text mode"),
     GNUNET_YES, &GNUNET_GETOPT_set_string, &dumpFileName},
    {'a', "address", NULL,
     gettext_noop ("address to send email to in case of email mode"),
     GNUNET_YES, &GNUNET_GETOPT_set_string, &emailAddress},
    {'d', "database", NULL, gettext_noop ("path to Expression Database file"),
     GNUNET_YES, &GNUNET_GETOPT_set_string, &edbFilePath},
    {'g', "gdb", NULL,
     gettext_noop ("path to gdb binary in use; default is /usr/bin/gdb"),
     GNUNET_YES, &GNUNET_GETOPT_set_string, &gdbBinaryPath},
    {'f', "function", NULL, gettext_noop ("Monkey will set a breakpoint on this function"),
     GNUNET_YES, &GNUNET_GETOPT_set_string, &inspectFunction},
    {'i', "inspect", NULL, gettext_noop ("An expression to inspect in the function specified after the argument f"),
     GNUNET_YES, &GNUNET_GETOPT_set_string, &inspectExpression},
    GNUNET_GETOPT_OPTION_END
  };

  if (argc < 2)
    {
      printf ("%s",
	      "Monkey should take arguments: Use --help to get a list of options.\n");
      return 1;
    }

  if (GNUNET_OK == GNUNET_PROGRAM_run (argc,
				       argv,
				       "gnunet-monkey",
				       gettext_noop
				       ("Automatically debug a service"),
				       options, &run, NULL))
    {
      return ret;
    }

  return 1;
}
